PCGen-Formula/code/src/java/pcgen/base/formula/visitor/SemanticsVisitor.java

```diff
@@ -336,46 +338,82 @@ public Object visit(ASTPCGenLookup node, Object data)
 		 * structures inside of it)
 		 */
 		ASTPCGenSingleWord ftnNode = (ASTPCGenSingleWord) firstChild;
-		String ftnName = ftnNode.getText();
+		String name = ftnNode.getText();
 		Node argNode = node.jjtGetChild(1);
-		Function function;
-		String functionForm;
-		FunctionLibrary library = fm.getLibrary();
+		Node[] args = VisitorUtilities.accumulateArguments(argNode);
 		if (argNode instanceof ASTFParen)
 		{
-			function = library.getFunction(ftnName);
-			functionForm = "()";
+			Function function = fm.getLibrary().getFunction(name);
+			if (function == null)
+			{
+				FormulaSemanticsUtilities.setInvalid(semantics, "Function: "
+					+ name + " was not found (called as: " + name + "(...))");
+				return semantics;
+			}
+			//Extract arguments from the grouping to give them to the function
+			function.allowArgs(this, args, semantics);
 		}
 		else if (argNode instanceof ASTPCGenBracket)
 		{
-			function = library.getBracketFunction(ftnName);
-			functionForm = "[]";
+			processArrayReference(name, args, semantics);
 		}
 		else
 		{
 			FormulaSemanticsUtilities.setInvalid(semantics,
 				"Parse Error: Function Formula Arguments"
 					+ " received invalid argument format, "
-					+ "expected: ASTFParen, got "
+					+ "expected: () or [], got "
 					+ firstChild.getClass().getName() + ": " + firstChild);
 			return semantics;
 		}
-		if (function == null)
+		return semantics;
+	}
+
+	private void processArrayReference(String name, Node[] args,
+		FormulaSemantics semantics)
+	{
+		if (args.length != 1)
 		{
-			FormulaSemanticsUtilities.setInvalid(semantics, "Function: "
-				+ ftnName + " was not found (called as: " + ftnName
-				+ functionForm + ")");
-			return semantics;
+			FormulaSemanticsUtilities.setInvalid(semantics, "Array: " + name
+				+ " requires only one (integer) argument, found: "
+				+ args.length);
+			return;
 		}
-		//Extract arguments from the grouping to give them to the function
-		int argLength = argNode.jjtGetNumChildren();
-		Node[] args = new Node[argLength];
-		for (int i = 0; i < argLength; i++)
+		Node arg = args[0];
+		if (!(arg instanceof ASTNum))
 		{
-			args[i] = argNode.jjtGetChild(i);
+			FormulaSemanticsUtilities.setInvalid(semantics,
+				"Array Reference requires an integer, "
+					+ "had invalid argument type: " + arg.getClass());
+			return;
```

Comment:
You are only supporting a single integer argument? The syntax supports more than this, why are the semantics limiting it?